src/main/java/seedu/partyplanet/model/person/Birthday.java

```diff
@@ -5,42 +5,131 @@
+import java.time.MonthDay;
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.TemporalAccessor;
 
 /** Represents a Person's birthday in PartyPlanet.
  * Guarantees: immutable; is always valid.
  */
-public class Birthday {
-
-    public static final String MESSAGE_CONSTRAINTS = "Birthday must be in the format yyyy-mm-dd";
-    public static final String MESSAGE_YEAR_CONSTRAINTS =
-            String.format("Year should not exceed %d", LocalDate.now().getYear());
+public class Birthday implements Comparable<Birthday> {
 
+    public static final String MESSAGE_CONSTRAINTS = "Birthdays should be in one of the following formats:\n"
+            + "    - yyyy-mm-dd (ISO format)\n"
+            + "    - dd.mm.yyyy\n"
+            + "    - dd/mm/yyyy\n"
+            + "    - dd mmm yyyy\n"
+            + "    - mmm dd yyyy";
+    public static final String MESSAGE_DATE_CONSTRAINTS = "Birthday should not be a date in the future";
     public static final String EMPTY_BIRTHDAY_STRING = "";
     public static final Birthday EMPTY_BIRTHDAY = new Birthday();
 
+    private static final DateTimeFormatter[] VALID_FORMATS = new DateTimeFormatter[] {
+            DateTimeFormatter.ofPattern("yyyy-MM-dd"),
+            DateTimeFormatter.ofPattern("d.M.yyyy"),
+            DateTimeFormatter.ofPattern("d/M/yyyy"),
+            DateTimeFormatter.ofPattern("d MMM yyyy"),
+            DateTimeFormatter.ofPattern("d MMMM yyyy"),
+            DateTimeFormatter.ofPattern("MMM d yyyy"),
+            DateTimeFormatter.ofPattern("MMMM d yyyy"),
+    };
+    private static final DateTimeFormatter[] VALID_FORMATS_WITHOUT_YEAR = new DateTimeFormatter[] {
+            DateTimeFormatter.ofPattern("--MM-dd"),
+            DateTimeFormatter.ofPattern("d/M"),
+            DateTimeFormatter.ofPattern("d MMM"),
+            DateTimeFormatter.ofPattern("d MMMM"),
+            DateTimeFormatter.ofPattern("MMM d"),
+            DateTimeFormatter.ofPattern("MMMM d"),
+    };
+    private static final DateTimeFormatter ISO_FORMAT = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+    private static final DateTimeFormatter ISO_FORMAT_WITHOUT_YEAR = DateTimeFormatter.ofPattern("--MM-dd");
+    private static final DateTimeFormatter READABLE_FORMAT = DateTimeFormatter.ofPattern("d MMM yyyy");
+    private static final DateTimeFormatter READABLE_FORMAT_WITHOUT_YEAR = DateTimeFormatter.ofPattern("d MMM");
+
+    /** In "dd mmm yyyy" formatted string for human-readable display */
+    public final String displayValue;
+    /** In ISO-8601 format for easy comparison */
     public final String value;
+    private int month; // implemented for quick retrieval
+    private boolean hasYear;
     private boolean isEmpty = false;
 
     /**
      * Constructs a {@code Birthday}.
+     * Birthdate should not be in the future, and can optionally contain a year.
+     * Some invalid dates are mapped to the nearest valid date, e.g. 29 Feb 2021 -> 28 Feb 2021.
```

Comment:
@zhengruoxin yup, also noted here. I couldn't find a way to solve this invalid date mapping problem, without having to move away from manual date parsing. Was planning to document this quirk in the UserGuide and leave it be, since adding the wrong date is not within normal use anyway.